Assets/Mirror/Runtime/QuaternionReadWrite.cs

```diff
@@ -0,0 +1,348 @@
+using UnityEngine;
+
+namespace Mirror
+{
+
+    /*
+        uncompressed Quaternion = 32 * 4 = 128 bits => send 16 bytes
+
+        Quaternion Always normalized
+        x^2 + y^2 + z^2 + w^2 = 1
+        can drop largest value and re-calculate it
+        encode which is largest using 2 bits
+
+        2nd largest value has max size of 1/sqrt(2)
+        c^2 + c^2 + 0 + 0 = 1
+        can encode the smallest three components in [-0.707107,+0.707107] instead of [-1,+1]
+
+
+        Sign of largest value doesnt matter
+        Q * vec3 == (-Q) * vec3
+
+
+        With no precision loss 
+
+        22 bits per value (32 * 0.707)
+        2 + 22 * 3 = 70 bits => send 9 bytes
+
+        9 bits per value
+        2 + 9 * 3 = 29 bits => send 4 bytes
+
+        Links for more info
+        https://youtu.be/Z9X4lysFr64
+        https://gafferongames.com/post/snapshot_compression/
+         */
+
+    public enum RotationPrecision
+    {
+        /// <summary>
+        /// Send 9 bytes, smallest rotation 0.000000334
+        /// </summary>
+        Full,
+        /// <summary>
+        /// Send 4 bytes, smallest rotation 0.00276
+        /// </summary>
+        Half,
+        /// <summary>
+        /// Dont send rotation
+        /// </summary>
+        NoRotation,
+    };
+    public static class QuaternionReadWrite
+    {
+        public static RotationPrecision DefaultPrecision => RotationPrecision.Half;
+        const float MinValue = -1f / 1.414214f; // 1/ sqrt(2)
+        const float MaxValue = 1f / 1.414214f;
+        const float ValueRange = MaxValue - MinValue;
+
+        /// <summary>
+        /// Writes Quaternion using QuaternionReadWrite.DefaultPrecision
+        /// </summary>
+        public static void WriteQuaternion(this NetworkWriter writer, Quaternion value)
```

Comment:
why not use WriteQuaternion with a default parameter? would get rid of the DefaultPrecision state above.

Reply:
I think `Codacy/PR Quality Review` shows an issue when using default parameter?

Reply:
using default parameter also breaks Weaver because of this check  

https://github.com/vis2k/Mirror/blob/fd0e2b18c30a5b4830df8c7a0c4a70eeb690effa/Assets/Mirror/Editor/Weaver/Processors/ReaderWriterProcessor.cs#L57